driver-core/src/test/functional/com/mongodb/client/model/expressions/ExpressionsFunctionalTest.java

```diff
@@ -0,0 +1,646 @@
+                CURRENT.getFieldDocumentArray("arrObj"),
+                "{'$getField': {'input': '$$CURRENT', 'field': 'arrObj'}}");
+
+        assertExpression(
+                BsonArray.parse("[true, false]"),
+                CURRENT.getFieldDocumentArray("arrObj")
+                        .map(d -> d.getFieldBoolean("v")),
+                " {'$map': {'input': {'$getField': {"
+                        + "'input': '$$CURRENT', 'field': 'arrObj'}}, "
+                        + "'in': {'$getField': {'input': '$$this', 'field': 'v'}}}}");
+    }
+
+    @Test
+    public void documentExpressionsTest() {
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/#object-expression-operators
+        // (Complete as of 6.0)
+
+        // literals
+        DocumentExpression a1 = Expressions.ofDocument(BsonDocument.parse("{a: 1}"));
+        DocumentExpression ax1ay2 = Expressions.ofDocument(BsonDocument.parse("{a: {x: 1, y: 2}}"));
+        DocumentExpression nullDoc = Expressions.ofDocument(null);
+        DocumentExpression emptyDoc = Expressions.ofDocument(BsonDocument.parse("{}"));
+        DocumentExpression literalDoc = Expressions.ofDocument(BsonDocument.parse("{lit: {'$not': true} }"));
+
+        assertExpression(BsonDocument.parse("{'a': 1}"), a1, "{'$literal': {'a': 1}}");
+        assertExpression(BsonDocument.parse("{'a': {'x': 1, 'y': 2}}"), ax1ay2, "{'$literal': {'a': {'x': 1, 'y': 2}}}");
+        assertExpression(null, nullDoc, "null");
+        assertExpression(BsonDocument.parse("{}"), emptyDoc, "{'$literal': {}}");
+        assertExpression(BsonDocument.parse("{'lit': {'$not': true}}"), literalDoc, "{'$literal': {'lit': {'$not': true}}}");
+        assertTrue(a1 instanceof DocumentExpression);
+
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/getField/ (100)
+        assertExpression(1, a1.getFieldInteger("a"), "{'$getField': {'input': {'$literal': {'a': 1}}, 'field': 'a'}}");
+        assertExpression(2, a1.getFieldInteger("a").multiply(2), "{'$multiply': [{'$getField': {'input': {'$literal': {'a': 1}}, 'field': 'a'}}, 2]}");
+
+        // missing fields
+        assertExpression(
+                BsonDocument.parse("{'a': 1}"),
+                a1.setField("z", a1.getFieldBoolean("missing")));
+        assertExpression(
+                BsonDocument.parse("{'a': 1}"),
+                a1.setField("z", a1.getFieldDocument("missing").getFieldDocument("also_missing")));
+        assertExpression(
+                BsonDocument.parse("{'a': 1, 'z': ''}"),
+                a1.setField("z", a1.getFieldString("missing").toLower()));
+        /*
+        The behaviour of missing fields appears to be as follows, and equivalent to $$REMOVE:
+        propagates -- getField, cond branches...
+        false -- not, or, cond check...
+        0     -- sum...
+        ""    -- toLower...
+        null  -- multiply, add, subtract, year, filter, reduce, map, result within map...
```

Comment:
I'm not sure what should be done with missing fields. They are like an optional (a document maps from strings to optionals). Imagine we have an array of products, some of which are missing the field, and the user writes the following (the `2` is some sort of surcharge):

```
products.map(p => p.getField("USD").sum(2)).reduce(sum)
```
Possible intended results:

1. An exception because any one product is missing the field. Alternatively, a null value (in place of an error).
2. For those products that have a USD, halve the value, and then sum the results. The field is optional, the mult [is mapped](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#map-java.util.function.Function-) on the optional, and the reduction pre-filters-out all empty optionals.

The current MQL/JSON API is not consistent in the approach it takes. Why does `toLower` return the empty string, allowing for a proper concat reduction, while array operations do not yield the empty array? If the same problem arises for all reductive operators (concat, union, or, and, add, mult, ...), why do we have sum instead of asking the user to place a filter prior to add? In the above code, if there are only 10 products, and all are missing the field, the result might not be 0 but 20.

Reply:
The user is able to use `getInteger(name, default)`. They can also filter out values. We may, in the future, want to consider providing `filterToInteger` methods on arrays, but this is uncertain.